### PR TITLE
Use pinning for generators to make trait safe

### PR DESCRIPTION
**Pull request by Nemo157**
I'm unsure whether there needs to be any changes to the actual generator transform. Tests are passing so the fact that `Pin<&mut T>` is fundamentally the same as `&mut T` seems to allow it to still work, but maybe there's something subtle here that could go wrong.

This is specified in [RFC 2349 § Immovable generators](https://github.com/rust-lang/rfcs/blob/master/text/2349-pin.md#immovable-generators) (although, since that RFC it has become safe to create an immovable generator, and instead it's unsafe to resume any generator; with these changes both are now safe and instead the unsafety is moved to creating a `Pin<&mut [static generator]>` which there are safe APIs for).

CC #43122

### CODE REVIEW

**Comment by Nemo157** [comment]
```diff
-	error[E0277]: the trait bound `[static generator@$DIR/static-not-unpin.rs:19:25: 21:6 _]: std::marker::Unpin` is not satisfied
+	error[E0277]: the trait bound `[static generator@$DIR/static-not-unpin.rs:19:25: 21:6 _]: std::pin::Unpin` is not satisfied
```

It definitely mentions `std::marker::Unpin` when I run this test locally (on `x86_64-apple-darwin`). Is the choice of which alias to use maybe not guaranteed?

**Comment by Nemo157** [comment]
So, that caused my new test case to pass, but broke a lot of other cases that relied on the previous ordering. I would have thought that the default ordering would be by declaration order, but I don't know how that could be different between my machines and travis'.

File: src/liballoc/boxed.rs [hunk 1]
```diff
@@ -804,13 +804,12 @@ impl<T: ?Sized> AsMut<T> for Box<T> {
...
 
 #[unstable(feature = "generator_trait", issue = "43122")]
-impl<T> Generator for Box<T>
-    where T: Generator + ?Sized
-{
-    type Yield = T::Yield;
-    type Return = T::Return;
-    unsafe fn resume(&mut self) -> GeneratorState<Self::Yield, Self::Return> {
-        (**self).resume()
+impl<G: ?Sized + Generator + Unpin> Generator for Box<G> {
```
Comment (withoutboats) [inline src/liballoc/boxed.rs]: Can you add an impl for `Pin<Box<G>>` also?

**Comment by withoutboats** [comment]
So this is basically what we always meant to do, just at the time Zoxc implemented the previous iteration of the generator API, `Pin` didn't exist yet. This will require a significant rewrite of futures-async-await, though my impression is that no one is using that and we maybe should just consider it completely deprecated instead.

std API changes look good. :+1:

@bors r? @eddyb can you review the compiler changes here and also confirm that the generator transform doesn't need to be changed with this PR

**Comment by eddyb** [comment]
The generator transform should use `*(self.0)` to refer to the state of type `Self`, instead of `*self` - I'm kind of surprised this doesn't trip anything already.

cc @Zoxc

**Comment by Nemo157** [comment]
I think I can see how to update the transform. At a guess it currently works because the output MIR is self-consistent with what type it thinks the generator argument is, and the type it thinks the argument is is layout compatible with the actual type passed in. There must not be any validation of the signature after the transform happens.

**Comment by Zoxc** [comment]
I suggesting replacing `self` with `self.0` in [create_generator_resume_function](https://github.com/rust-lang/rust/blob/master/src/librustc_mir/transform/generator.rs#L743) as we still use just `self` in the destructor.

**Comment by Nemo157** [comment]
I think I have to also update the debug-info generation: https://github.com/rust-lang/rust/blob/ddd4b194a0c28065a277c1f964aa931cc021b5c6/src/librustc_codegen_llvm/mir/mod.rs#L593-L598

I have a patch to do that, but I still have to figure out how to test it. There appear to be some debuginfo tests, but nothing covering generators, and when I run the existing tests I get

```
During startup program terminated with signal SIG113, Real-time event 113.
```

File: src/librustc_codegen_llvm/mir/mod.rs [hunk 2]
```diff
@@ -590,10 +590,15 @@ fn arg_local_refs(
                 return;
             }
 
+            let pin_did = tcx.lang_items().pin_type().unwrap();
```
Comment (eddyb) [inline src/librustc_codegen_llvm/mir/mod.rs]: Please keep this in `Option` form and use `Some(def.did) == pin_did` below.

File: src/test/ui/generator/static-not-unpin.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// normalize-stderr-test "std::pin::Unpin" -> "std::marker::Unpin"
```
Comment (Nemo157) [inline src/test/ui/generator/static-not-unpin.rs]: I have not been able to figure out why travis shows a different path for this compared to the two machines with rustc I've been able to test on. I attempted to trace through where the diagnostics get the path from, and as far as I can tell everything should be deterministic. The path chosen appears to come from the [`visible_parent_map` built here](https://github.com/rust-lang/rust/blob/653da4fd006c97625247acd7e076d0782cdc149b/src/librustc_metadata/cstore_impl.rs#L314), this is done as a BFS, and should be deterministic within a single crate. Following the order of an items children back from here through the different representations all the way to `libsyntax` there doesn't appear to be anything that should change the order, then in `libsyntax` an items children come out in parse order. `std` re-exports the `core::marker` module before the `core::pin` module, so the path chosen to be displayed should be `std::marker::Pin`.

Parse/declaration order is also consistent with the failures in https://travis-ci.org/rust-lang/rust/jobs/451369080, that test commit forced the child order while constructing `visible_parent_map` to be sorted by name. All the failures were swapping paths that were declared earlier with paths declared later but alphabetically sorted earlier.

This will also become moot soon as https://github.com/rust-lang/rust/issues/55766 is proposing to remove the `std::pin::Unpin` re-export (although, it does plan to add it to the prelude which will then become the preferred path).

**Comment by Nemo157** [comment]
Travis appears to have found all the test cases I didn't...

Other than the comment I just left this is complete as far as I'm aware. I'd prefer to land this with that workaround and I'll open an issue about the non-deterministic diagnostics. Let me know whether there's anything else and I can squash all the fixups in.

**Comment by Nemo157** [comment]
@Zoxc yep, [here's one MIR dump]https://gist.github.com/Nemo157/ecd934261d28e59c013a1499c37ce00d) I had laying around still. Once I changed the generator transform itself I actually got an assertion during codegen because the signature didn't match, that's how I found the signature being generated [here](https://github.com/rust-lang/rust/pull/55704/files#diff-44a4a8b18aebfe4cd2261ccc019a1f0fR89) and updated that as well. So I'm fairly confident it's all lining up correctly now.

File: src/libstd/future.rs [hunk 4]
```diff
@@ -43,7 +43,9 @@ impl<T: Generator<Yield = ()>> !Unpin for GenFuture<T> {}
 impl<T: Generator<Yield = ()>> Future for GenFuture<T> {
     type Output = T::Return;
     fn poll(self: Pin<&mut Self>, lw: &LocalWaker) -> Poll<Self::Output> {
-        set_task_waker(lw, || match unsafe { Pin::get_mut_unchecked(self).0.resume() } {
+        // Safe because we're !Unpin + !Drop mapping to a ?Unpin value
+        let gen = unsafe { Pin::map_unchecked_mut(self, |s| &mut s.0) };
+        set_task_waker(lw, || match gen.resume() {
```
Comment (Zoxc) [inline src/libstd/future.rs]: @withoutboats Did you look at this?

File: src/test/run-pass/generator/auxiliary/xcrate-reachable.rs [hunk 5]
```diff
@@ -8,15 +8,16 @@
...
 
+use std::marker::Unpin;
 use std::ops::Generator;
 
 fn msg() -> u32 {
     0
 }
 
-pub fn foo() -> impl Generator<Yield=(), Return=u32> {
+pub fn foo() -> impl Generator<Yield=(), Return=u32> + Unpin {
```
Comment (Zoxc) [inline src/test/run-pass/generator/auxiliary/xcrate-reachable.rs]: Why did you add the `Unpin` bound here and on other tests?

File: src/test/run-pass/generator/auxiliary/xcrate.rs [hunk 6]
```diff
@@ -8,19 +8,20 @@
...
+pub fn foo() -> impl Generator<Yield = (), Return = ()> + Unpin {
     || {
         if false {
             yield;
         }
     }
 }
 
-pub fn bar<T: 'static>(t: T) -> Box<Generator<Yield = T, Return = ()>> {
+pub fn bar<T: Unpin + 'static>(t: T) -> Box<Generator<Yield = T, Return = ()> + Unpin> {
```
Comment (Zoxc) [inline src/test/run-pass/generator/auxiliary/xcrate.rs]: The `Unpin` bound on `T` here is strange too.

File: src/librustc/traits/select.rs [hunk 7]
```diff
@@ -1671,6 +1671,19 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                 self.assemble_builtin_bound_candidates(clone_conditions, &mut candidates)?;
             }
 
+            if lang_items.unpin_trait() == Some(def_id) {
+                // OK to skip binder because the substs on generator types never
+                // touch bound regions, they just capture the in-scope
+                // type/region parameters
```
[the remark this answers is not in the document]
Comment (Nemo157) [reply src/librustc/traits/select.rs]: Updated comment.

That's what I was thinking of initially, but I couldn't find anything similar to that (if you have some pointers for where to look I can take another look at it). This definitely [gives the correct error message](https://github.com/rust-lang/rust/pull/55704/files#diff-d745a2cbea2137958fc6da09c5e6006aR1).

File: src/test/run-pass/generator/auxiliary/xcrate-reachable.rs (hunk 5 above)
Comment (Nemo157) [reply src/test/run-pass/generator/auxiliary/xcrate-reachable.rs]: `Unpin` is necessary to use `Pin::new`, otherwise the tests would need to use the unsafe `Pin::new_unchecked` (this may not actually be necessary with `impl Trait`, `Unpin` probably leaks like other auto-traits, I didn't check that).

File: src/test/run-pass/generator/auxiliary/xcrate.rs (hunk 6 above)
Comment (Nemo157) [reply src/test/run-pass/generator/auxiliary/xcrate.rs]: I think that's necessary because non-`static` generators are still subject to the normal structural `Unpin` checks. Rebuilding without it now to check (will be a while as I've got to do a fresh build).

File: src/librustc/traits/select.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Nemo157) [reply src/librustc/traits/select.rs]: Actually, I think I've found a better place to put this. In `assemble_candidates_from_auto_impls` I can just suppress adding a candidate for static generators when the trait is `Unpin`. I'll check whether that works now.

File: src/test/run-pass/generator/static-generators.rs [hunk 8]
```diff
@@ -21,8 +22,7 @@ fn main() {
         yield;
         assert_eq!(b as *const _, &a as *const _);
     };
-    unsafe {
-        assert_eq!(generator.resume(), GeneratorState::Yielded(()));
-        assert_eq!(generator.resume(), GeneratorState::Complete(()));
-    }
+    let mut generator = unsafe { Pin::new_unchecked(&mut generator) };
```
Comment (Zoxc) [inline src/test/run-pass/generator/static-generators.rs]: Why does this use unsafe code?

File: src/test/run-pass/generator/auxiliary/xcrate-reachable.rs (hunk 5 above)
Comment (Zoxc) [reply src/test/run-pass/generator/auxiliary/xcrate-reachable.rs]: Yeah, I was expecting `Unpin` to leak here, so this change shouldn't be necessary.

File: src/librustc/traits/select.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Zoxc) [reply src/librustc/traits/select.rs]: @nikomatsakis @arielb1 Where would the correct place to put this be? Or does it not matter since the user can't write impls since generators are unnameable?

File: src/test/run-pass/generator/static-generators.rs (hunk 8 above)
Comment (Nemo157) [reply src/test/run-pass/generator/static-generators.rs]: Constructing a `Pin` to a reference to `!Unpin` value is unsafe. There is a safe [macro-based solution](https://docs.rs/pin-utils/0.1.0-alpha.3/pin_utils/macro.pin_mut.html) to do so on the stack, but it's not included in `std`. I could use a `Pin<Box>` here instead safely, but I thought better to just keep it on the stack.

I can add a safety note about why this is safe (same reason as `pin_mut!` is safe, it shadows the owning binding so the value cannot be moved after this line).

File: src/test/run-pass/generator/auxiliary/xcrate.rs (hunk 6 above)
Comment (Nemo157) [reply src/test/run-pass/generator/auxiliary/xcrate.rs]: So, I haven't spent time verifying this, but having non-`static` generators use the structural definition of pinning makes this sound code:

```rust
fn foo() -> impl Generator<Yield = usize, Return = ()> {
    static || {
        let x = 5;
        let x = &x;
        yield *x;
        yield *x;
    }
}

fn bar() -> impl Generator<Yield = usize, Return = ()> {
    || {
        let mut foo = foo();
        loop {
            let val = {
                let mut foo = unsafe { Pin::new_unchecked(&mut foo) };
                match foo.resume() {
                    GeneratorState::Yielded(val) => val,
                    GeneratorState::Complete(()) => break,
                }
            };
            yield val;
        }
    }
}
```

I don't know if there would be a way to make this _safe_ code, the stack pinning macro would require borrows to persist across yields, but maybe something like this would be useful?

It also seems to me that this should be investigated for both closures and generators at the same time, they likely have similar reasons to either propagate or force `Unpin`.

@withoutboats any opinion (and does closures interaction with `Unpin` matter for stabilization of `Pin`)?

Comment (withoutboats) [reply src/test/run-pass/generator/auxiliary/xcrate.rs]: There should be some way to make generators which are checked for `Unpin`, so they can be moved in and out of pins (and thus moved between resume calls). Currently its not annotating the generator `static`.

I think the current closure behavior around `Unpin` is correct in that it is consistent with other auto traits like `Send` etc; if anything should change (not sure that it should) it would be generator syntax.

Comment (Nemo157) [reply src/test/run-pass/generator/auxiliary/xcrate.rs]: > maybe it only should check that they don't themselves borrow across yields, and omit the generated negative impl of unpin.

That's basically what is happening now, for `static` annotated generators there's an implicit negative impl of `Unpin`, for other generators there's no implicit impl (either positive or negative) and it's falling back to structural checking of the upvars and state that is live across yield points (the closed over _environment_).

We could force a positive impl of `Unpin` for non-`static` generators instead and say that it's unsound to create a pinned reference into the environment, even just between yield points. You would still be able to create pinned references to values that are only live between yield points since they are true stack variables. And if you need pinned references into the environment you can force the generator to be immovable via `static`. As far as I'm aware this wouldn't remove any capabilities; a non-`static` generator that is `!Unpin` because of structural recursion and a `static` generator appear the same externally, and when implementing the generator you always have the choice to add or remove the `static` annotation

---

Thinking about it a bit more I guess this doesn't matter for closure's as we don't have `FnPin`, there's no way to call a closure such that it could soundly create a pinned reference into its environment.

File: src/test/run-pass/generator/auxiliary/xcrate-reachable.rs (hunk 5 above)
Comment (Nemo157) [reply src/test/run-pass/generator/auxiliary/xcrate-reachable.rs]: Removed the leaking `Unpin`s.

File: src/test/run-pass/generator/static-generators.rs (hunk 8 above)
Comment (Nemo157) [reply src/test/run-pass/generator/static-generators.rs]: Added a note on the safety here.

File: src/test/run-pass/generator/auxiliary/xcrate.rs (hunk 6 above)
Comment (Nemo157) [reply src/test/run-pass/generator/auxiliary/xcrate.rs]: I've added an unconditional implementation of `Unpin` for movable generators.

**Comment by Dylan-DPC** [comment]
ping from triage
@Zoxc awaiting your review on this
@Nemo157 can you update this by resolving the conflicts?

**Comment by Nemo157** [comment]
Pushed an untested rebase, there were only trivial import conflicts in the `run-pass` tests (that didn't even renumber the error logs) so shouldn't have broken anything.

**Comment by inkyu-prestolabs** [comment]
Can I ask a question? previously, i was storing a generator using this way:
```
pub struct MyStruct {
    pub gen: Box<dyn std::ops::Generator<Return = Result<(), Error>, Yield = MyYield>>,
}
```

But after this change, how can I store a generator in a struct?

**Comment by Arnavion** [comment]
```
pub struct MyStruct {
    pub gen: std::pin::Pin<Box<dyn std::ops::Generator<Return = Result<(), Error>, Yield = MyYield>>>,
}
```

which you can get from `Box::pin`
